engine/access/rpc/backend/connection_factory.go

```diff
@@ -74,24 +89,53 @@ func (cf *ConnectionFactoryImpl) createConnection(address string, timeout time.D
 
 func (cf *ConnectionFactoryImpl) retrieveConnection(grpcAddress string, timeout time.Duration) (*grpc.ClientConn, error) {
 	var conn *grpc.ClientConn
+	var clientMutex *sync.Mutex
+	var store *ConnectionCacheStore
+
+	cf.mutex.Lock()
 	if res, ok := cf.ConnectionsCache.Get(grpcAddress); ok {
-		conn = res.(*grpc.ClientConn)
+		store = res.(*ConnectionCacheStore)
+		clientMutex = store.mutex
+		conn = store.ClientConn
 		if cf.AccessMetrics != nil {
 			cf.AccessMetrics.ConnectionFromPoolRetrieved()
 		}
+	} else {
+		clientMutex = new(sync.Mutex)
+		store = &ConnectionCacheStore{
+			ClientConn: nil,
+			mutex:      clientMutex,
+		}
+		cf.ConnectionsCache.Add(grpcAddress, store)
 	}
-	if conn == nil || conn.GetState() != connectivity.Ready {
-		// This line ensures that when a connection is renewed, the previously cached connection is evicted and closed
-		cf.ConnectionsCache.Remove(grpcAddress)
+	clientMutex.Lock()
+	defer clientMutex.Unlock()
+	cf.mutex.Unlock()
+
+	if conn == nil || conn.GetState() == connectivity.Shutdown {
+		// updates to the cache don't trigger evictions; this line closes connections before re-establishing new ones
+		if conn != nil {
+			conn.Close()
+		}
 		var err error
-		conn, err = cf.createConnection(grpcAddress, timeout)
+		conn, err = cf.addConnection(grpcAddress, timeout, store)
 		if err != nil {
 			return nil, err
 		}
-		cf.ConnectionsCache.Add(grpcAddress, conn)
-		if cf.AccessMetrics != nil {
-			cf.AccessMetrics.TotalConnectionsInPool(uint(cf.ConnectionsCache.Len()), cf.CacheSize)
-		}
+	}
+	return conn, nil
+}
+
+func (cf *ConnectionFactoryImpl) addConnection(grpcAddress string, timeout time.Duration, store *ConnectionCacheStore) (*grpc.ClientConn, error) {
+	conn, err := cf.createConnection(grpcAddress, timeout)
+	if err != nil {
+		return nil, err
+	}
+
+	store.ClientConn = conn
+
+	if cf.AccessMetrics != nil {
+		cf.AccessMetrics.TotalConnectionsInPool(uint(cf.ConnectionsCache.Len()), cf.CacheSize)
```

Comment:
since we're in here anyway. any other metrics we could add? Maybe a counter for new connections. it probably makes sense to add a debug level log for that as well as for evictions.